File: cathode-jobs/src/main/java/net/simonvt/cathode/remote/sync/movies/SyncAnticipatedMovies.java
/*
 * Copyright (C) 2016 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.remote.sync.movies;

import android.content.ContentProviderOperation;
import android.content.ContentResolver;
import android.content.ContentValues;
import android.database.Cursor;
import java.util.ArrayList;
import java.util.List;
import javax.inject.Inject;
import net.simonvt.cathode.api.entity.AnticipatedItem;
import net.simonvt.cathode.api.entity.Movie;
import net.simonvt.cathode.api.enumeration.Extended;
import net.simonvt.cathode.api.service.MoviesService;
import net.simonvt.cathode.jobqueue.JobPriority;
import net.simonvt.cathode.provider.DatabaseContract.MovieColumns;
import net.simonvt.cathode.provider.MovieDatabaseHelper;
import net.simonvt.cathode.provider.ProviderSchematic.Movies;
import net.simonvt.cathode.remote.CallJob;
import net.simonvt.schematic.Cursors;
import retrofit2.Call;

public class SyncAnticipatedMovies extends CallJob<List<AnticipatedItem>> {

  private static final int LIMIT = 50;

  @Inject transient MoviesService moviesService;

  @Inject transient MovieDatabaseHelper movieHelper;

  @Override public String key() {
    return "SyncAnticipatedMovies";
  }

  @Override public int getPriority() {
    return JobPriority.SUGGESTIONS;
  }

  @Override public Call<List<AnticipatedItem>> getCall() {
    return moviesService.getAnticipatedMovies(LIMIT, Extended.FULL);
  }

  @Override public boolean handleResponse(List<AnticipatedItem> movies) {
    ContentResolver resolver = getContentResolver();

    ArrayList<ContentProviderOperation> ops = new ArrayList<>();
    Cursor c = resolver.query(Movies.ANTICIPATED, new String[] {
        MovieColumns.ID,
    }, null, null, null);
    List<Long> movieIds = new ArrayList<>();
    while (c.moveToNext()) {
      final long movieId = Cursors.getLong(c, MovieColumns.ID);
      movieIds.add(movieId);
    }
    c.close();

    for (int i = 0, count = movies.size(); i < count; i++) {
      AnticipatedItem item = movies.get(i);
      Movie movie = item.getMovie();

      final long movieId = movieHelper.partialUpdate(movie);

      movieIds.remove(movieId);

      ContentValues cv = new ContentValues();
      cv.put(MovieColumns.ANTICIPATED_INDEX, i);
      ContentProviderOperation op =
          ContentProviderOperation.newUpdate(Movies.withId(movieId)).withValues(cv).build();
      ops.add(op);
    }

    for (Long movieId : movieIds) {
      ContentValues cv = new ContentValues();
      cv.put(MovieColumns.ANTICIPATED_INDEX, -1);
      ContentProviderOperation op =
          ContentProviderOperation.newUpdate(Movies.withId(movieId)).withValues(cv).build();
      ops.add(op);
    }

    return applyBatch(ops);
  }
}

File: cathode-sync/src/main/java/net/simonvt/cathode/actions/movies/SyncAnticipatedMovies.kt
/*
 * Copyright (C) 2016 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.actions.movies

import android.content.ContentProviderOperation
import android.content.ContentValues
import android.content.Context
import net.simonvt.cathode.actions.CallAction
import net.simonvt.cathode.api.entity.AnticipatedItem
import net.simonvt.cathode.api.enumeration.Extended
import net.simonvt.cathode.api.service.MoviesService
import net.simonvt.cathode.common.database.forEach
import net.simonvt.cathode.common.database.getLong
import net.simonvt.cathode.provider.DatabaseContract.MovieColumns
import net.simonvt.cathode.provider.ProviderSchematic.Movies
import net.simonvt.cathode.provider.batch
import net.simonvt.cathode.provider.helper.MovieDatabaseHelper
import net.simonvt.cathode.provider.query
import net.simonvt.cathode.settings.SuggestionsTimestamps
import retrofit2.Call
import javax.inject.Inject

class SyncAnticipatedMovies @Inject constructor(
  private val context: Context,
  private val movieHelper: MovieDatabaseHelper,
  private val moviesService: MoviesService
) : CallAction<Unit, List<AnticipatedItem>>() {

  override fun key(params: Unit): String = "SyncAnticipatedMovies"

  override fun getCall(params: Unit): Call<List<AnticipatedItem>> =
    moviesService.getAnticipatedMovies(LIMIT, Extended.FULL)

  override suspend fun handleResponse(params: Unit, response: List<AnticipatedItem>) {
    val ops = arrayListOf<ContentProviderOperation>()
    val movieIds = mutableListOf<Long>()

    val localMovies = context.contentResolver.query(Movies.ANTICIPATED, arrayOf(MovieColumns.ID))
    localMovies.forEach { cursor -> movieIds.add(cursor.getLong(MovieColumns.ID)) }
    localMovies.close()

    response.forEachIndexed { index, anticipatedItem ->
      val movie = anticipatedItem.movie!!
      val movieId = movieHelper.partialUpdate(movie)
      movieIds.remove(movieId)

      val values = ContentValues()
      values.put(MovieColumns.ANTICIPATED_INDEX, index)
      val op = ContentProviderOperation.newUpdate(Movies.withId(movieId)).withValues(values).build()
      ops.add(op)
    }

    for (movieId in movieIds) {
      val values = ContentValues()
      values.put(MovieColumns.ANTICIPATED_INDEX, -1)
      val op = ContentProviderOperation.newUpdate(Movies.withId(movieId)).withValues(values).build()
      ops.add(op)
    }

    context.contentResolver.batch(ops)

    SuggestionsTimestamps.get(context)
      .edit()
      .putLong(SuggestionsTimestamps.MOVIES_ANTICIPATED, System.currentTimeMillis())
      .apply()
  }

  companion object {
    private const val LIMIT = 50
  }
}
